### ISSUE TITLE
Unusually high row reads

### ISSUE THREAD

**Issue by heymartinadams** [prisma/prisma#10993]
### Bug description

This issue can probably be largely ignored because it’s likely an edge case, but wanted to post it here for anyone else running into a similar issue.

A month ago, I noticed extremely high row reads in our [PlanetScale database](https://planetscale.com) — over 1 billion row reads over a period of 30 days (our plan only allotted us 100M row reads per month). For a [small app](https://LifePurposeApp.com) with a database of ~56K, these many row reads pointed to an error somewhere.

### How to reproduce

It turns out that a nested `some: {}` Prisma query was responsible for the excess. I had implemented the following schema that accounted for a `User` as well as a `UserAlias` table since some users used multiple email addresses to sign in, and we wanted to make sure that whichever email they used would get them to the same account:

```prisma
model User {
  UserAliases      UserAlias[] @relation("UserAliasesToOUser")
  UserAlias        UserAlias?  @relation("OUserAliasToOUser")
  email            String      @unique
}

model UserAlias {
  id          String   @id @default(uuid())
  User        User?    @relation("UserAliasesToOUser", fields: [userEmail], references: [email])
  userEmail   String? // User’s default email
  Alias       User?    @relation("OUserAliasToOUser", fields: [aliasEmail], references: [email])
  aliasEmail  String?  @unique

  @@index([aliasEmail])
  @@index([userEmail])
}
```

Each time a user logged in, the following Prisma query was run:

<img width="829" alt="prisma1" src="https://user-images.githubusercontent.com/11673745/148435196-7e911c1e-cd0d-4343-853d-6218b11a5ea7.png">

The purpose of this was to get the user’s information, regardless of which email they used.

*It turns out that doing a nested query using `some: {}` effectively scans the entire table.* So each time a user logged in, all 56K records where read, and this soon ended up being over a billion rows being read each month.

The fix itself was easy: I just broke up the `where` query into two separate Prisma reads:

<img width="694" alt="prisma2" src="https://user-images.githubusercontent.com/11673745/148435762-515d82ee-0334-4e75-b6b4-8743fc7d8285.png">

### Prisma information

See above.

### Environment & setup

- OS: macOS
- Database: MySQL using PlanetScale
- Node.js version: v16.10.0

### Prisma Version

```
v3.4.1
```

**Comment by janpio** [comment]
Thanks for the report, that is indeed interesting to know.

If you have the time, can you enable query logging and post the SQL query that Prisma generated and executed under the hood for each of the queries? This might make it clearer what is going on.

I assume `prismaUserGetOne` is just `prisma.user.findUnique()` under the hood?

**Comment by pantharshit00** [comment]
Yeah, if you can share `prismaUserGetOne` function, that would be great.

**Comment by heymartinadams** [comment]
Sure! @janpio the SQL query that caused the issue was the following:
```sql
select `User`.id, `User`.email
from `User` 
where `User`.id in (select t0.id from `User` as t0 join UserAlias as j0 on j0.userEmail = t0.email where j0.id in (select t1.id from UserAlias as t1 join `User` as j1 on j1.email = t1.aliasEmail where j1.email = 'example@email.com' and t1.id is not null) and t0.id is not null) or `User`.email = 'example@email.com' 
order by `User`.createdAt asc limit 0, 1
```

Indeed, @pantharshit00 I abstract Prisma operations into resolver files.

<img width="624" alt="CleanShot 2022-01-07 at 06 20 56@2x" src="https://user-images.githubusercontent.com/11673745/148557172-c0d6f4c5-ed8f-478e-8e77-4ea205c3f729.png">

**Comment by heymartinadams** [comment]
@janpio I’ve been using `prisma.user.findFirst()` instead of `findUnique()` since that allows for `AND/OR` operators inside the `where` query (see https://github.com/prisma/prisma/issues/6440).

**Comment by pantharshit00** [comment]
Thanks for the details. I can also confirm that `some` is causing the read upsurge here. We should review its implementation.

**Comment by janpio** [comment]
Without taking a deeper look on the code @heymartinadams , do you think a missing index could be involved here, so another variant of the problem described in https://github.com/prisma/prisma/issues/10611?

**Comment by heymartinadams** [comment]
Sorry my delay @janpio. I’ve definitely been including indices (`@@index([someId])`) according to recommended guidelines.

**Comment by heymartinadams** [comment]
Looks like PlanetScale implemented a pricing structure that can accommodate higher row reads! I believe this was at least partially done in response to this particular issue. In my other app, [Ecstatic](https://ecstatic.com), we use a lot of `some: {...}` queries. I was afraid I’d be running into the same issue once we launched.

https://planetscale.com/blog/increasing-planetscale-plan-limits-to-include-billions-of-reads

**Comment by jaivinwylde** [comment]
I'm having the same issue with findFirst and planetscale. The following query was only used around 350 times, but it created over 6.5 million row reads.
Query:
```
          const user = await ctx.database.user.findFirst({
            where: {
              [key]: value,
            },
            select: {
              id: true,
              createdAt: true,
              username: true,
              email: true,
              emailVerified: true,
            },
          })
```
Schema:
```
model User {
  id        Int       @id @default(autoincrement())
  createdAt DateTime  @default(now())
  expiresAt DateTime?

  userUid         String  @default(cuid())
  email           String  @db.VarChar(255)
  emailVerified   Boolean @default(false)
  username        String  @db.VarChar(16)
  displayUsername String  @db.VarChar(16)
  name            String? @db.VarChar(32)

  @@unique([userUid, email, username], name: "identifiers")
  @@index([userUid, email, username])
}
```
This is obviously way too many row reads. Our setup looks similar to @heymartinadams solution, but we still have this problem. We would prefer findUnique, but we have 3 uniques per User (username, email, uid) and only know 1 at the time of the request (username or email), and I don't know how we could use findUnique with only one unique.
